sql: Support unnest(anyarray, anyarray, anyarray...)

giorgosp opened the pull request:
Previously `unnest` supported only one array argument. This commit
adds a variadic overload in order to support multiple array arguments,
as in Postgres.

The variadic overload is supported in a SELECT clause as well, in contrast
with Postgres that only supports it only in a FROM clause.

This change has two side effects.

1. The first side effect is that tuples are allowed to be declared with
duplicate labels, like:

`SELECT ((1, '2') AS a, a);`

This statement executes successfully since the tuple labels are declared
but not used.

However, accessing a duplicate tuple label returns AmbiguousColumn error.
Both statements below return a "column reference <column> is ambiguous"
error:

`SELECT (((1,2,3) AS a,b,a)).a;`
`SELECT ((unnest(ARRAY[1,2], ARRAY[1,2]))).unnest;`

2. The `to_json` builtin doesn't return duplicate error anymore but
succeeds when passed in values with duplicate labels. Eg:

```sql
SELECT to_json(x.*) FROM (VALUES (1,2)) AS x(column2);
+----------------+
|    to_json     |
|----------------|
| {"column2": 2} |
+----------------+
```

This is a bit more compatible with Postgres which also doesn't return
an error for this query. However, there is still an incompatibility in
the results as Postgres renders the duplicate labels as duplicate json
keys. The correct test cases have been added as a TODO comment, tracked
by #44465.

Release note (sql change): An overload was added to the `unnest` builtin
function in order to support multiple array arguments.

Release note (sql change): Duplicate labels are allowed when declaring a
tuple, but an Ambiguous column error is returned if a duplicate
label is accessed. For example:
`SELECT ((1, '2') AS a, a);` is successful but
`SELECT (((1,2,3) AS a,b,a)).a;` returns an error.

Fixes #41434

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41557)
<!-- Reviewable:end -->

giorgosp commented:
Hi @jordanlewis ! I have created an overload of `unnest` that supports variadic arguments. However I have some issues and would like some advice on how to proceed:

The labels are hardcoded when passed to `genProps` and it seems the `FunctionProperties. ReturnLabels` take precedence over the labels returned by the `ReturnTyper` or the `multipleArrayValueGenerator.ResolvedType()`.

If I pass `nil` to `genProps`, the variadic overload works but not the simple overload. I could try adding a new `makeUnnestBuiltin` that would set the labels appropriately and create the overloads inside itself, except if you have a better idea to keep the code consistent (using the already provided workflow with `makeBuiltin()`/`genProps()`/`makeGeneratorOverload()`) so that the overloads are easily visible and not created deeper in the stack.

This is the exact (trimmed) error from
`make testlogic TESTCONFIG=local FILES='srfs' SUBTESTS='unnest'`:

```
SELECT unnest(ARRAY[1,2], ARRAY['a','b']);
--- PASS: TestExecBuild (1.20s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestExecBuild113095743
    test_log_scope.go:58: use -show-logs to present logs inline
    --- PASS: TestExecBuild/local (0.28s)
        --- PASS: TestExecBuild/local/srfs (0.28s)
testing: warning: no tests to run
PASS
panic: invalid datum type given: int, expected tuple{int AS unnest, string AS unnest} [recovered]
	panic: panic while executing 1 statements: SELECT unnest(ARRAY[_, _], ARRAY[_, _]); caused by invalid datum type given: int, expected tuple{int AS unnest, string AS unnest}

goroutine 1174 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00122c000, 0x3d1a620, 0xc000a95500, 0x3064460, 0xc0016f16f0)
	/home/gio/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:731 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc00122c000, 0x3d1a620, 0xc000a95500)
	/home/gio/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0x61
panic(0x3064460, 0xc0016f16f0)
	/usr/lib/go-1.12/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DatumToEncDatum(0xc001706a00, 0x3d876a0, 0xc0010afec8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/gio/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:167 +0x1b7
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*projectSetProcessor).toEncDatum(0xc001343400, 0x3d876a0, 0xc0010afec8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/gio/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/project_set.go:280 +0x9b
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*projectSetProcessor).nextGeneratorValues(0xc001343400, 0xc00100aa20, 0x1, 0x1)
	/home/gio/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/project_set.go:180 +0x340
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*projectSetProcessor).Next(0xc001343400, 0x0, 0x0, 0x0, 0x3d4cd60)
	/home/gio/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/project_set.go:259 +0x89
github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run(0x3d1a620, 0xc001205e00, 0x3d38f20, 0xc001343400, 0x3cff960, 0xc001423c00)
	/home/gio/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:169 +0x35
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run(0xc001343400, 0x3d1a620, 0xc001205e00)
	/home/gio/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:725 +0x92
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc001287560, 0x3d1a6e0, 0xc00100a990, 0x364c628, 0x0, 0x0)
	/home/gio/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:375 +0x1ec
```
Thank you!

yuzefovich commented:
Hi @giorgosp, thanks for taking this on! I looked into the relevant code, and it seems to me that current implementation of `makeBuiltin` does not support the case when two overloads have different `FunctionProperties` - and this is exactly the case that we have here with `unnest`. I'm actually not sure on how to properly address it - it seems to me that your idea of `makeUnnestBuiltin` is not sufficient because we have `generators` as `map[string]builtinDefinition`, and we need a map from `string + argument types` to `builtinDefinition`, so for `unnest` we need to be able to return a different builtin (with different props) depending on the input argument types.

Possibly another way to solve it is to modify the way SRFs are transformed `sql/transform_srf.go`, but I don't have a feeling whether it is doable/reasonable.

cc @jordanlewis @knz am I missing something?

knz commented:
This would need a redesign of how function properties work. 

If you wish to do this I think it would be better to extend the function properties mechanism (and the overload resolution code that goes with it) in a separate PR first, then come back to the current PR once the other PR is merged.

jordanlewis commented:
@giorgosp, considering that the known use case we're trying to support uses exactly 2 arguments, would it be possible to just add that one for now? I know that is an unsatisfying answer...

giorgosp commented:
Thank you for your comments! @jordanlewis If I understand correctly, you suggest we support `unnest(array, array)` only. 

My main driver for having two overloads was that Postgres supports `unnest(array,array)` only in in a FROM clause, but it seems like CockroachDB doesn't have that limitation.

So we could just keep the variadic one, which also won't break the current usages of unnest with one argument. What do you think?

jordanlewis commented:
No, I meant adding a second overload with 2 `anyarray` arguments, alongside the one with a single `anyarray` argument. I was under the impression from this conversation that it would be difficult to make variadic `unnest` work properly.

giorgosp commented:
As @yuzefovich and @knz said, the mechamism around `FunctionProperties` doesn't allow different function properties per overload. In our case we need different `FunctionProperties.ReturnLabels`. 

`unnest(anyarray)` shouldn't return a tuple so that `select unnest(ARRAY[1,2]) + 1` works. But
`unnest(anyarray...)` must return a tuple.

My suggestion to keep only the variadic overload won't work either.
I am still looking into it.

yuzefovich commented:
I agree with @giorgosp - I also think without making changes that Raphael pointed out it is not possible to add `unnest(anyarray,anyarray)` because it requires two labels, but currently a single `"unnest"` label is hardcoded in the function properties.

giorgosp commented:
After a second look, I found out it would be simple to completely remove the ReturnLabels which were hardcoded in FunctionProperties and have some flexibility as to what return labels will be used each time.

Opened a PR (#41861) to await for your feedback.

giorgosp committed:
sql: support unnest(anyelement[], anyelement[], anyelement[]...)

Previously `unnest` supported only one array argument. This commit
adds a variadic overload in order to support multiple array arguments,
as in Postgres.

The variadic overload is supported in a SELECT clause as well, in contrast
with Postgres that only supports it only in a FROM clause.

This change has two side effects.

1. The first side effect is that tuples are allowed to be declared with
duplicate labels, like:

`SELECT ((1, '2') AS a, a);`

This statement executes successfully since the tuple labels are declared
but not used.

However, accessing a duplicate tuple label returns AmbiguousColumn error.
Both statements below return a "column reference <column> is ambiguous"
error:

`SELECT (((1,2,3) AS a,b,a)).a;`
`SELECT ((unnest(ARRAY[1,2], ARRAY[1,2]))).unnest;`

2. The `to_json` builtin doesn't return duplicate error anymore but
succeeds when passed in values with duplicate labels. Eg:

```sql
SELECT to_json(x.*) FROM (VALUES (1,2)) AS x(column2);
+----------------+
|    to_json     |
|----------------|
| {"column2": 2} |
+----------------+
```

This is a bit more compatible with Postgres which also doesn't return
an error for this query. However, there is still an incompatibility in
the results as Postgres renders the duplicate labels as duplicate json
keys. The correct test cases have been added as a TODO comment, tracked
by #44465.

Release note (sql change): An overload was added to the `unnest` builtin
function in order to support multiple array arguments.

Release note (sql change): Duplicate labels are allowed when declaring a
tuple, but an Ambiguous column error is returned if a duplicate
label is accessed. For example:
`SELECT ((1, '2') AS a, a);` is successful but
`SELECT (((1,2,3) AS a,b,a)).a;` returns an error.

Fixes #41434

yuzefovich reviewed (commented):
Very nice work! I think you're very close to the finish line :)

> Reviewed 2 of 3 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-:-Lx4IlNI5LCon13V_tF2:bw7ssgy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @giorgosp and @jordanlewis)

___
*[pkg/sql/logictest/testdata/logic_test/srfs, line 1113 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4Jkwv5epyoPQHXi1Y:-Lx4Jkwv5epyoPQHXi1Z:b5exxgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/logictest/testdata/logic_test/srfs#L1113)):*
> ```
> 
> query T
> SELECT unnest(ARRAY[1,2], ARRAY['a'], ARRAY[1.1, 2.2, 3.3])
> ```

Hm, postgres (11.5 version) doesn't seem to support such case (of tuples), only `SELECT * FROM unnest...` is supported. I expected that there would be NULLs in the tuples when the corresponding array is "exhausted", so I wanted to check with PSQL. Shouldn't there be NULLs in this query?
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4HxNe2Nw9mcn9Wky4:-Lx4HxNfCkTtRphn9y7Y:b5nhhcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/sem/builtins/generator_builtins.go#L185)):*
> ```Go
> 					}
> 					returnTypes[i] = *arg.ResolvedType().ArrayContents()
> 					labels[i] = "unnest"
> ```

The build is complaining about duplicated labels, probably we'll need to include the counter into the label.
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 508 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4IR_M-1eS1Iaclihi:-Lx4IR_M-1eS1Iaclihj:bk0227v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/sem/builtins/generator_builtins.go#L508)):*
> ```Go
> 	for i, arr := range s.arrays {
> 		returnTypes[i] = *arr.ParamTyp
> 		labels[i] = "unnest"
> ```

ditto for the counter
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 535 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4IWSr8nyfRaqCJdvI:-Lx4IWSr8nyfRaqCJdvJ:b-romyzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/sem/builtins/generator_builtins.go#L535)):*
> ```Go
> // Values implements the tree.ValueGenerator interface.
> func (s *multipleArrayValueGenerator) Values() tree.Datums {
> 	var datums tree.Datums
> ```

[nit]: it probably makes sense to allocate the memory right away so that no allocations occur during `append`s below. We could take it one step further and perform the allocation in `Start` method, store the slice within the struct and reuse it on every call to `Values`.


<!-- Sent from Reviewable.io -->

giorgosp reviewed (commented):
Thanks! Please bare with me (some more) because I will be on vacation for a week. But afterwards I will work on it so we finish it as soon as possible.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-:-Lx8A0cRDTlvzkANzOst:b-n1j637)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @giorgosp, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/srfs, line 1113 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4Jkwv5epyoPQHXi1Y:-Lx8AJJ40w858KK8vh56:b-bwntvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/logictest/testdata/logic_test/srfs#L1113)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, postgres (11.5 version) doesn't seem to support such case (of tuples), only `SELECT * FROM unnest...` is supported. I expected that there would be NULLs in the tuples when the corresponding array is "exhausted", so I wanted to check with PSQL. Shouldn't there be NULLs in this query?
</blockquote></details>

Indeed postgres doesn't support `unnest` in a SELECT clause. NULLs are returned by the srf anyway but I guess cockroachdb doesn't render them? If `unnest` is used in a FROM clause, the NULLs are rendered.
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4HxNe2Nw9mcn9Wky4:-Lx8Ajzj2oKsqg5EJ3-S:bjweb7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/sem/builtins/generator_builtins.go#L185)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

The build is complaining about duplicated labels, probably we'll need to include the counter into the label.
</blockquote></details>

Thanks for checking it out! I wanted to ask for help about this. Do we really want the counter-in-label solution? Because in postgres all the labels have the same name. 

What is that duplicate labels check used for? Would it make sense to skip it for system-defined labels and keep it only for user defined ones (when AS is used in the query)? Postgres already has this quirk for unnest and `SELECT unnest FROM unnest(...)` returns "ambiguous column" error as someone would expect.


<!-- Sent from Reviewable.io -->

giorgosp reviewed (commented):
> Reviewed 1 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41557)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @giorgosp, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/sem/builtins/generator_builtins.go, line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4HxNe2Nw9mcn9Wky4:-Lz8L5mbCOYj6nho4zHC:bqz7h0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/sem/builtins/generator_builtins.go#L185)):*
<details><summary><i>Previously, giorgosp (George Papadrosou) wrote…</i></summary><blockquote>

Thanks for checking it out! I wanted to ask for help about this. Do we really want the counter-in-label solution? Because in postgres all the labels have the same name. 

What is that duplicate labels check used for? Would it make sense to skip it for system-defined labels and keep it only for user defined ones (when AS is used in the query)? Postgres already has this quirk for unnest and `SELECT unnest FROM unnest(...)` returns "ambiguous column" error as someone would expect.
</blockquote></details>

Still thinking about this. Adding counters to the unnest labels would finalize this PR but would create an incompatibility with postgres. Trying to figure out `func (expr *Tuple) TypeCheck` is not called with local config
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 535 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4IWSr8nyfRaqCJdvI:-Lz8Ayq19y2Gg-GEJgY-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/sem/builtins/generator_builtins.go#L535)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: it probably makes sense to allocate the memory right away so that no allocations occur during `append`s below. We could take it one step further and perform the allocation in `Start` method, store the slice within the struct and reuse it on every call to `Values`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41557)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @giorgosp, @jordanlewis, @knz, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/srfs, line 1113 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4Jkwv5epyoPQHXi1Y:-LzA_CziAHfOMXVxjQ1v:bgo2wp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/logictest/testdata/logic_test/srfs#L1113)):*
<details><summary><i>Previously, giorgosp (George Papadrosou) wrote…</i></summary><blockquote>

Indeed postgres doesn't support `unnest` in a SELECT clause. NULLs are returned by the srf anyway but I guess cockroachdb doesn't render them? If `unnest` is used in a FROM clause, the NULLs are rendered.
</blockquote></details>

Ok, sounds good to me.
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4HxNe2Nw9mcn9Wky4:-LzAbPTN0oINyHkdKLhN:b-77h22c) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/sem/builtins/generator_builtins.go#L185)):*
<details><summary><i>Previously, giorgosp (George Papadrosou) wrote…</i></summary><blockquote>

Still thinking about this. Adding counters to the unnest labels would finalize this PR but would create an incompatibility with postgres. Trying to figure out `func (expr *Tuple) TypeCheck` is not called with local config
</blockquote></details>

Hm, I see that in #24832 we purposefully decided to deviate from Postgres, so I think we now have two choices:
1. include the counter into the label in this PR. This will increase deviation from psql, but no other changes are needed
2. remove the requirement that the tuple's labels are unique. Then the counter is not needed, but we might need to change the way we access an element of the tuple by the label (if there are multiple matches, just use the first one, or something similar).

I don't know which one is preferable. Maybe @knz has thoughts?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 1 of 4 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41557)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @giorgosp, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/sem/builtins/generator_builtins.go, line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4HxNe2Nw9mcn9Wky4:-LzBQmDB-SD2R73oIwNR:bs9s0y0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/sem/builtins/generator_builtins.go#L185)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, I see that in #24832 we purposefully decided to deviate from Postgres, so I think we now have two choices:
1. include the counter into the label in this PR. This will increase deviation from psql, but no other changes are needed
2. remove the requirement that the tuple's labels are unique. Then the counter is not needed, but we might need to change the way we access an element of the tuple by the label (if there are multiple matches, just use the first one, or something similar).

I don't know which one is preferable. Maybe @knz has thoughts?
</blockquote></details>

I vote to remove the duplicate check in labels but at the same time do the same as pg and return an ambiguous error during the per-label lookup if the searched label is found two times.

(This will need dedicated unit tests)


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Very nice work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Could you please update the PR description message with the latest commit message? Also, an addition to release note is needed now that we allow duplicate labels but will error out on accessing them if it is ambigous.

> Reviewed 8 of 8 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-:-Lzi9P015DM5YWGejFqN:b9194ay)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @giorgosp, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/json_builtins, line 327 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lzi8BBs6EaPQH756FqN:-Lzi8BBs6EaPQH756FqO:b-4bd445) ([raw file](https://github.com/cockroachdb/cockroach/blob/c893d72fd069aff76a89167ce18bfecc096fc4cd/pkg/sql/logictest/testdata/logic_test/json_builtins#L327)):*
> ```
> {"a": 1, "column2": 2}
> 
> # TODO: Implement the test cases below to be compatible with Postgres
> ```

Do you know if we have an issue about this (probably not)? If not, could you please create one and mention it in this comment?
___
*[pkg/sql/logictest/testdata/logic_test/tuple, line 678 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lzi8eyt19g10_Q0wZRF:-Lzi8eyt19g10_Q0wZRG:b-cyrygw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c893d72fd069aff76a89167ce18bfecc096fc4cd/pkg/sql/logictest/testdata/logic_test/tuple#L678)):*
> ```
> (1,2,hello,,)  (t,,"(f,6.6,f)")
> 
> # Duplicate tuple labels are allowed (but access fails a duplicated label is accessed, see
> ```

[nit]: `s/access fails a/access fails when a/g` or something like that.
___
*[pkg/sql/sem/tree/type_check.go, line 610 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lzi9FL1CtfraKBKTae9:-Lzi9FL20ICANyVvnLbd:b-aw7gvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c893d72fd069aff76a89167ce18bfecc096fc4cd/pkg/sql/sem/tree/type_check.go#L610)):*
> ```Go
> 		if label == expr.ColName {
> 			if expr.ColIndex != -1 {
> 				// Found a duplicate label
> ```

[nit]: missing period.


<!-- Sent from Reviewable.io -->

giorgosp reviewed (commented):
Thank you! Updated the PR and the commit msg.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-:-LziJHYn9JrOXugGtHOy:b-pg68w9)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/json_builtins, line 327 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lzi8BBs6EaPQH756FqN:-LziHM2I5StfP45K3QW8:b-896fja) ([raw file](https://github.com/cockroachdb/cockroach/blob/c893d72fd069aff76a89167ce18bfecc096fc4cd/pkg/sql/logictest/testdata/logic_test/json_builtins#L327)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Do you know if we have an issue about this (probably not)? If not, could you please create one and mention it in this comment?
</blockquote></details>

Done!
___
*[pkg/sql/logictest/testdata/logic_test/tuple, line 678 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lzi8eyt19g10_Q0wZRF:-LziH_LW5s2FJe6ocrqI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c893d72fd069aff76a89167ce18bfecc096fc4cd/pkg/sql/logictest/testdata/logic_test/tuple#L678)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: `s/access fails a/access fails when a/g` or something like that.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lx4HxNe2Nw9mcn9Wky4:-LziHfYy6Dc2TXKjg3CB:bzn6td) ([raw file](https://github.com/cockroachdb/cockroach/blob/a281de641a9d59a4c917f367973f54173d2db139/pkg/sql/sem/builtins/generator_builtins.go#L185)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I vote to remove the duplicate check in labels but at the same time do the same as pg and return an ambiguous error during the per-label lookup if the searched label is found two times.

(This will need dedicated unit tests)
</blockquote></details>

Thank you all for your comments! I was about to add a comment but @yuzefovich beat me to it :)
___
*[pkg/sql/sem/tree/type_check.go, line 610 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41557#-Lzi9FL1CtfraKBKTae9:-LziD59kDTdOzpcqznoj:bg10q22) ([raw file](https://github.com/cockroachdb/cockroach/blob/c893d72fd069aff76a89167ce18bfecc096fc4cd/pkg/sql/sem/tree/type_check.go#L610)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: missing period.
</blockquote></details>

Right. I was expecting the linter to show an error btw :/


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Thanks for your contribution!

bors r=yuzefovich
